apps/api_web/test/api_web/params_test.exs

```diff
@@ -58,25 +58,34 @@ defmodule ApiWeb.ParamsTest do
     end
 
     test "filters items from 2nd arg", %{params: params} do
-      assert Params.filter_params(params, ["route", "trip"]) == %{
-               "route" => "1,2,3",
-               "trip" => "1,2,3"
-             }
+      assert Params.filter_params(params, ["route", "trip", "stop"]) ==
+               {:ok,
+                %{
+                  "route" => "1,2,3",
+                  "trip" => "1,2,3",
+                  "stop" => "4,5,6"
+                }}
     end
 
-    test "takes query params in the `filter` namespace", %{params: params} do
-      assert Params.filter_params(params, ["route", "stop"]) == %{
-               "route" => "1,2,3",
-               "stop" => "4,5,6"
-             }
+    test "returns error in case of unsupported filter", %{params: params} do
+      assert Params.filter_params(params, ["route"]) == {:error, :bad_filter, ~w(stop trip)}
+    end
+  end
+
+  describe "validate_includes/2" do
+    test "returns ok for valid includes" do
+      assert Params.validate_includes(%{"include" => "stops,trips"}, ~w(stops routes trips)) ==
+               {:ok, ~w(stops trips)}
     end
 
-    test "params in `filter` namespace have priority over duplicate params", %{params: params} do
-      assert Params.filter_params(params, ["stop"]) == %{"stop" => "4,5,6"}
+    test "returns error for invalid includes" do
+      assert Params.validate_includes(%{"include" => "stops,routes"}, ~w(stops trips)) ==
+               {:error, :bad_include, ~w(routes)}
     end
 
-    test "invalid filter values are removed" do
-      assert Params.filter_params(%{"filter" => "string"}, ["stop"]) == %{}
+    test "supports dot notation" do
+      assert Params.validate_includes(%{"include" => "stops.id"}, ~w(stops routes trips)) ==
```

Comment:
shouldn't this be `{:error, :bad_include, ~w(id)}` ?

Reply:
Maybe it's more trouble than it's worth, though.

Reply:
Yes, I thought that validation of the top level part (that is, everything before the dot) should generally be enough in terms of maintaining value vs complexity balance. Let me know if you want me to look into the implementation of additional validations.

Reply:
No, I think this should be good for now.